services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadPartRequestPublisher.java

```diff
@@ -0,0 +1,379 @@
+    private final long lastPartSize;
+
+    private final Object lock = new Object();
+
+    public UploadPartRequestPublisher(Builder builder) {
+        this.uploadId = builder.uploadId;
+        this.optimalPartSize = builder.partSize;
+        this.putObjectRequest = builder.putObjectRequest;
+        this.asyncRequestBody = builder.asyncRequestBody;
+        this.byteBufferStoringSubscriber = new ByteBufferStoringSubscriber(builder.partSize);
+        this.numOfPartsBuffered = builder.numOfPartsBuffered;
+        this.executor = builder.executor;
+        int totalParts = builder.numOfParts;
+        numOfRemainingParts = new AtomicInteger(totalParts);
+        lastPartSize = builder.contentLength - optimalPartSize * (totalParts - 1);
+    }
+
+    public static Builder builder() {
+        return new UploadPartRequestPublisher.Builder();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super Pair<UploadPartRequest, AsyncRequestBody>> subscriber) {
+        subscriber.onSubscribe(new RequestSubscription(subscriber));
+    }
+
+    private class RequestSubscription implements Subscription {
+        private final Subscriber<? super Pair<UploadPartRequest, AsyncRequestBody>> subscriber;
+
+        RequestSubscription(Subscriber<? super Pair<UploadPartRequest, AsyncRequestBody>> subscriber) {
+            this.subscriber = subscriber;
+        }
+
+        @Override
+        public void request(long n) {
+            log.trace(() -> "received request, current demand: " + outstandingDemand.get());
+            if (n <= 0) {
+                signalSubscriberError(n);
+                return;
+            }
+
+            if (done) {
+                return;
+            }
+
+            outstandingDemand.updateAndGet(current -> {
+                if (Long.MAX_VALUE - current < n) {
+                    return Long.MAX_VALUE;
+                }
+
+                return current + n;
+            });
+
+
+            if (subscribed.compareAndSet(false, true)) {
+                asyncRequestBody.subscribe(byteBufferStoringSubscriber);
+            }
+
+            if (requests.isEmpty()) {
+                prepareRequests();
+            }
+            flushBuffer();
+        }
+
+        private void signalSubscriberError(long n) {
+            subscriber.onSubscribe(new NoopSubscription(subscriber));
+            IllegalArgumentException failure = new IllegalArgumentException("A downstream publisher requested an invalid "
+                                                                            + "amount of data: " + n);
+            subscriber.onError(failure);
+        }
+
+        @Override
+        public void cancel() {
+            // TODO: cancel upstream
+        }
+
+        public void flushBuffer() {
+            if (isFlushingBuffer.compareAndSet(false, true)) {
+                try {
+                    log.trace(() -> "Flushing buffer, current outstanding demand: " + outstandingDemand.get());
```

Comment:
What if multiple threads have updated the `outstandingDemand` value at line 113 before calling `flushBuffer()` here?